File: src/services/MetricsService/MetricsJsonSerializer.ts
import {Metric} from './Metric';

export interface MetricsJsonSerializer {
  serializeToJson(timestamp: number, metrics: Metric[]): string;
}

export class DefaultMetricsJsonSerializer implements MetricsJsonSerializer {
  private appVersion: string;
  private appOs: string;
  private osVersion: string;
  private manufacturer: string;
  private model: string;
  private androidReleaseVersion: string;

  constructor(
    appVersion: string,
    appOs: string,
    osVersion: string,
    manufacturer: string,
    model: string,
    androidReleaseVersion: string,
  ) {
    this.appVersion = appVersion;
    this.appOs = appOs;
    this.osVersion = osVersion;
    this.manufacturer = manufacturer;
    this.model = model;
    this.androidReleaseVersion = androidReleaseVersion;
  }

  serializeToJson(timestamp: number, metrics: Metric[]): string {
    function serializeDynamicPayload(map: [string, string][]): any {
      return Array.from(map).reduce((acc, [key, value]) => {
        // @ts-ignore
        acc[key] = value;
        return acc;
      }, {});
    }

    const jsonStructure = {
      metricstimestamp: timestamp,
      appversion: this.appVersion,
      appos: this.appOs,
      osversion: this.osVersion,
      manufacturer: this.manufacturer,
      model: this.model,
      androidreleaseversion: this.androidReleaseVersion,
      payload: metrics.map(metric => {
        return {
          ...{identifier: metric.identifier, region: metric.region, timestamp: metric.timestamp},
          ...serializeDynamicPayload(metric.payload),
        };
      }),
    };
    return JSON.stringify(jsonStructure);
  }
}

File: android/app/src/main/java/app/covidshield/services/metrics/MetricsJsonSerializer.kt
package app.covidshield.services.metrics

import com.google.gson.Gson

interface MetricsJsonSerializer {
    fun serializeToJson(timestamp: Number, metrics: List<Metric>): String
}

class DefaultMetricsJsonSerializer(
        private val appVersion: String,
        private val appOs: String,
        private val osVersion: String,
        private val manufacturer: String,
        private val model: String,
        private val androidReleaseVersion: String) : MetricsJsonSerializer {

    override fun serializeToJson(timestamp: Number, metrics: List<Metric>): String {
        val jsonStructure = mapOf<String, Any>(
                "metricstimestamp" to timestamp,
                "appversion" to appVersion,
                "appos" to appOs,
                "osversion" to osVersion,
                "manufacturer" to manufacturer,
                "model" to model,
                "androidreleaseversion" to androidReleaseVersion,
                "payload" to metrics.map { metric ->
                    mapOf<String, Any>(
                            "identifier" to metric.identifier,
                            "region" to metric.region,
                            "timestamp" to metric.timestamp
                    ) + metric.payload
                }
        )
        return Gson().toJson(jsonStructure)
    }

}